### PR TITLE
dde.gradients support 3D outputs

### PR DESCRIPTION
**Pull request by Jerry-Jzy**

### CODE REVIEW

**Comment by lululxvi** [comment]
This PR should only modify dde.gradients, not pde_operator.py.

**Comment by lululxvi** [comment]
You need to test all backends.

**Comment by Jerry-Jzy** [comment]
> You need to test all backends.

The only case where the output is 3D is shown to be supported only by tensorflow v2 and pytorch

https://github.com/lululxvi/deepxde/blob/master/examples/operator/stokes_aligned_pideeponet.py

**Comment by lululxvi** [comment]
Not necessary testing PI-DeepONet. You can manually construct a function, and jus test if dde.gradients works correctly.

**Comment by Jerry-Jzy** [comment]
> Not necessary testing PI-DeepONet. You can manually construct a function, and jus test if dde.gradients works correctly.

I have tested pytorch, tf v2, paddle, jax. I don't have access to the environment for tf v1

File: deepxde/gradients/gradients.py [hunk 1]
```diff
@@ -19,7 +19,7 @@ def jacobian(ys, xs, i=None, j=None):
       computation.
 
     Args:
-        ys: Output Tensor of shape (batch_size, dim_y).
+        ys: Output Tensor of shape (batch_size, dim_y) or (batch_size1, batch_size2, dim_y).
```
Comment (lululxvi) [inline deepxde/gradients/gradients.py:22]: `xs` is (batch_size, dim_x), so is this one (batch_size_cond, batch_size, dim_y) or (batch_size, batch_size_cond, dim_y)? `batch_size_cond` means additional batch_size

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients.py:22]: here is  (batch_size_cond, batch_size, dim_y). in PI-DeepONet, `batch_size_cond` is the batch size of source term in branch net input

Comment (lululxvi) [reply deepxde/gradients/gradients.py:22]:
```
ys: Output Tensor of shape (batch_size, dim_y) or (batch_size_out, batch_size, dim_y). Here, the `batch_size` is the same one for `xs`, and `batch_size_out` is the batch size for an additional dimension.
```

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients.py]: ok

File: deepxde/gradients/gradients.py [hunk 2]
```diff
@@ -19,7 +19,9 @@ def jacobian(ys, xs, i=None, j=None):
       computation.
 
     Args:
-        ys: Output Tensor of shape (batch_size, dim_y).
+        ys: Output Tensor of shape (batch_size, dim_y) or (batch_size_out, batch_size, dim_y).
```
Comment (lululxvi) [inline deepxde/gradients/gradients.py:22]: max 88 characters per line

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients.py]: ok

Comment (lululxvi) [reply deepxde/gradients/gradients.py]: Also need to modify the doc of `hessian`

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients.py]: sure

File: deepxde/gradients/gradients.py [hunk 3]
```diff
@@ -19,7 +19,9 @@ def jacobian(ys, xs, i=None, j=None):
       computation.
 
     Args:
-        ys: Output Tensor of shape (batch_size, dim_y).
+        ys: Output Tensor of shape (batch_size, dim_y) or (batch_size_out, batch_size,
+            dim_y). Here, the `batch_size` is the same one for `xs`, and
+            `batch_size_out` is the batch size for an additional dimension.
```
Comment (lululxvi) [inline deepxde/gradients/gradients.py:24]:
```
`batch_size_out` is the batch size for an additional/outer dimension.
```

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients.py]: Ok

File: deepxde/gradients/gradients.py [hunk 4]
```diff
@@ -19,7 +19,9 @@ def jacobian(ys, xs, i=None, j=None):
       computation.
 
     Args:
-        ys: Output Tensor of shape (batch_size, dim_y).
+        ys: Output Tensor of shape (batch_size, dim_y) or (batch_size_out, batch_size,
+            dim_y). Here, the `batch_size` is the same one for `xs`, and
+            `batch_size_out` is the batch size for an additional/outer dimension.
```
Comment (lululxvi) [inline deepxde/gradients/gradients.py:24]: Add the dimension in line 33 and 62.

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients.py:24]: Ok

File: deepxde/gradients/gradients.py [hunk 5]
```diff
@@ -28,7 +30,9 @@ def jacobian(ys, xs, i=None, j=None):
             unless J has only one element, which is returned.
 
     Returns:
-        (`i`, `j`)th entry J[`i`, `j`], `i`th row J[`i`, :], or `j`th column J[:, `j`].
+        (`i`, `j`)th entry J[`i`, `j`], `i`th row J[`i`, :], or `j`th column J[:, `j`],
+        with 2 dimension when ys of shape (batch_size, dim_y) or 3 dimension when
+        ys of shape (batch_size_out, batch_size, dim_y).
```
Comment (lululxvi) [inline deepxde/gradients/gradients.py:35]:
```
(`i`, `j`)th entry J[`i`, `j`], `i`th row J[`i`, :], or `j`th column J[:, `j`].
When `ys` has shape (batch_size, dim_y), the output shape is ... When `ys` has shape (...), ...
```

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients.py]: I have updated the comments. This comment is based on the assumption that 3D output is already supported by pde_operator.py

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients.py]: If we treat the module as an independent module when writing the comments, then the `ys` shape can be something else like `(batch_size_out, batch_size)` or `(batch_size, batch_size_out)`.

Comment (lululxvi) [reply deepxde/gradients/gradients.py]: We have already designed the rule that it is (batch_size_out, batch_size, dim_y) in line 22.

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients.py]: Ok, I understand. Then, reverse-mode and forward-mode need to be considered separately

File: deepxde/gradients/gradients.py [hunk 6]
```diff
@@ -29,6 +31,10 @@ def jacobian(ys, xs, i=None, j=None):
 
     Returns:
         (`i`, `j`)th entry J[`i`, `j`], `i`th row J[`i`, :], or `j`th column J[:, `j`].
+        When `ys` has shape (batch_size, dim_y), the output shape is (batch_size, 1).
+        When `ys` has shape (batch_size_out, batch_size, dim_y), the output shape is
+        (batch_size_out, batch_size, 1) if forward-mode autodiff is used or
+        (batch_size, 1) if reverse-mode autodiff is used.
```
[the remark this answers is not in the document]
Comment (Jerry-Jzy) [reply deepxde/gradients/gradients.py:37]: reverse mode does support batch_size_out. In reverse mode, the gradient has the same shape as the input since it uses vector jacobian product. so the shape is (batch_size, 1) no matter reverse_mode or forward-mode is used.

[the remark this answers is not in the document]
Comment (lululxvi) [reply deepxde/gradients/gradients.py:37]: Then in the comment of `ys`, you need to add that `(batch_size_out, batch_size, dim_y)` is only supported by forward-mode. Also, add an error checking: if it is reverse mode with batch_size_out, then raise an error.

[the remark this answers is not in the document]
Comment (lululxvi) [reply deepxde/gradients/gradients.py:37]: I don't understand. If revise-mode supports batch_size_out, then why the output is (batch_size, 1)? Where is batch_size_out?

[the remark this answers is not in the document]
Comment (Jerry-Jzy) [reply deepxde/gradients/gradients.py:37]: reverse-mode support `ys` shape has the shape (batch_size_out, batch_size, dim_y) and `xs` has the shape (batch_size, dim_x). That's this PR care about, 3D output.

File: deepxde/gradients/gradients_reverse.py [hunk 7]
```diff
@@ -21,11 +21,11 @@ def __call__(self, i=None, j=None):
         # Compute J[i, :]
         if i not in self.J:
             if backend_name in ["tensorflow.compat.v1", "tensorflow"]:
-                y = self.ys[:, i : i + 1] if self.dim_y > 1 else self.ys
+                y = self.ys[..., i : i + 1] if self.dim_y > 1 else self.ys
```
Comment (lululxvi) [inline deepxde/gradients/gradients_reverse.py:24]: Since reverse mode doesn't support 3D anyway, we don't need to modify this file.

File: deepxde/gradients/gradients.py [hunk 8]
```diff
@@ -29,8 +32,15 @@ def jacobian(ys, xs, i=None, j=None):
 
     Returns:
         (`i`, `j`)th entry J[`i`, `j`], `i`th row J[`i`, :], or `j`th column J[:, `j`].
+        When `ys` has shape (batch_size, dim_y), the output shape is (batch_size, 1).
+        When `ys` has shape (batch_size_out, batch_size, dim_y), the output shape is
+        (batch_size_out, batch_size, 1).
     """
     if config.autodiff == "reverse":
+        if bkd.ndim(ys) == 3:
```
Comment (lululxvi) [inline deepxde/gradients/gradients.py:40]: Raise this error in gradients_reverse.py

File: deepxde/gradients/gradients_reverse.py (hunk 7 above)
Comment (Jerry-Jzy) [reply deepxde/gradients/gradients_reverse.py]: Ok

File: deepxde/gradients/gradients.py (hunk 8 above)
Comment (Jerry-Jzy) [reply deepxde/gradients/gradients.py]: Sure

File: deepxde/gradients/jacobian.py [hunk 9]
```diff
@@ -20,22 +20,22 @@ def __init__(self, ys, xs):
         self.xs = xs
 
         if backend_name in ["tensorflow.compat.v1", "paddle"]:
-            self.dim_y = ys.shape[1]
+            self.dim_y = ys.shape[-1]
```
[the remark this answers is not in the document]
Comment (Jerry-Jzy) [reply deepxde/gradients/jacobian.py:25]: Ok

File: deepxde/gradients/jacobian.py [hunk 10]
```diff
@@ -114,7 +116,9 @@ def __call__(self, ys, xs, i=None, j=None):
         #     x = torch.from_numpy(x)
         #     x.requires_grad_()
         #     f(x)
-        if backend_name in ["tensorflow.compat.v1", "tensorflow"]:
+        if backend_name == "tensorflow.compat.v1":
```
Comment (lululxvi) [inline deepxde/gradients/jacobian.py:119]: Why this change?

Comment (Jerry-Jzy) [reply deepxde/gradients/jacobian.py:119]: Since tensorflow.compat.v1 and tensorflow have different `ys`. treat them separately is more clear

Comment (lululxvi) [reply deepxde/gradients/jacobian.py:119]: The original code is equivalent to
```
if backend_name == "tensorflow.compat.v1":
    if config.autodiff == "reverse":
        key = (ys.ref(), xs.ref())
    elif config.autodiff == "forward":
        key = (ys[0].ref(), xs.ref())
elif backend_name == "tensorflow":
    if config.autodiff == "reverse":
        key = (ys.ref(), xs.ref())
    elif config.autodiff == "forward":
        key = (ys[0].ref(), xs.ref())
```
So you mean that for tensorflow.compat.v1, `key = (ys.ref(), xs.ref())` works for both reverse and forward?

Comment (lululxvi) [reply deepxde/gradients/jacobian.py:119]: Oh, so you say the original code has a bug?

Comment (Jerry-Jzy) [reply deepxde/gradients/jacobian.py:119]: Yes

File: deepxde/gradients/gradients_reverse.py [hunk 11]
```diff
@@ -4,6 +4,7 @@
 
 from .jacobian import Jacobian, Jacobians
 from ..backend import backend_name, tf, torch, jax, paddle
+from .. import backend as bkd
```
Comment (lululxvi) [inline deepxde/gradients/gradients_reverse.py:7]:
```
from .. import backend as bkd
from ..backend import backend_name, tf, torch, jax, paddle
```

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients_reverse.py:6]: Ok

File: deepxde/gradients/gradients_reverse.py [hunk 12]
```diff
@@ -70,6 +71,10 @@ def __call__(self, i=None, j=None):
 
 
 def jacobian(ys, xs, i=None, j=None):
+    if bkd.ndim(ys) == 3:
```
Comment (lululxvi) [inline deepxde/gradients/gradients_reverse.py:74]: Move this error message to `JacobianReverse`

File: deepxde/gradients/gradients_reverse.py [hunk 13]
```diff
@@ -137,6 +142,10 @@ def clear(self):
 
 
 def hessian(ys, xs, component=0, i=0, j=0):
+    if bkd.ndim(ys) == 3:
```
Comment (lululxvi) [inline deepxde/gradients/gradients_reverse.py:145]: Move this error checking to `Hessian`

File: deepxde/gradients/gradients_reverse.py (hunk 12 above)
Comment (Jerry-Jzy) [reply deepxde/gradients/gradients_reverse.py]: Ok

File: deepxde/gradients/gradients_reverse.py (hunk 13 above)
Comment (Jerry-Jzy) [reply deepxde/gradients/gradients_reverse.py]: Ok

File: deepxde/gradients/gradients_reverse.py [hunk 14]
```diff
@@ -91,15 +100,20 @@ class Hessian:
...
+            ndim_y = bkd.ndim(ys[0])
         if component >= dim_y:
             raise ValueError(
                 "The component of ys={} cannot be larger than the dimension={}.".format(
                     component, dim_y
                 )
             )
-
+        if ndim_y == 3:
+            raise NotImplementedError(
```
Comment (lululxvi) [inline deepxde/gradients/gradients_reverse.py:114]: This is not needed, as it will go to Jacobian, which raises the error.

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients_reverse.py]: Ok

File: deepxde/gradients/gradients_reverse.py [hunk 15]
```diff
@@ -17,6 +18,14 @@ def __call__(self, i=None, j=None):
                     "Reverse-mode autodiff doesn't support computing a column."
                 )
             i = 0
+        if backend_name in ["tensorflow.compat.v1", "tensorflow", "pytorch", "paddle"]:
+            ndim_y = bkd.ndim(ys)
```
Comment (lululxvi) [inline deepxde/gradients/gradients_reverse.py:22]: self.ys

Comment (Jerry-Jzy) [reply deepxde/gradients/gradients_reverse.py]: fixed
